static/js/topic_generator.js

```diff
@@ -2,197 +2,48 @@
-                        blueslip.error("Invalid generator returned.");
-                        return undefined;
-                    }
-                }
-                const inner = inner_gen.next();
-                if (inner !== undefined) {
-                    return inner;
-                }
-                outer_val = outer_gen.next();
-                inner_gen = undefined;
+        for (let i = 0; i < curr_topic_index; i += 1) {
+            const topic = topics[i];
+            if (has_unread_messages(stream, topic)) {
+                return {stream, topic};
             }
-            return undefined;
-        },
-    };
-};
-
-exports.chain = function (gen_lst) {
-    function get(which) {
-        return which;
-    }
-
-    const outer_gen = exports.list_generator(gen_lst);
-
-    return exports.fchain(outer_gen, get);
-};
-
-exports.wrap = function (lst, val) {
-    if (val === undefined) {
-        return exports.list_generator(lst);
-    }
-
-    const i = lst.indexOf(val);
-    if (i < 0) {
-        return exports.list_generator(lst);
-    }
-
-    const inners = [
-        exports.sub_list_generator(lst, i, lst.length),
-        exports.sub_list_generator(lst, 0, i),
-    ];
-
-    return exports.chain(inners);
-};
-
-exports.wrap_exclude = function (lst, val) {
-    if (val === undefined) {
-        return exports.list_generator(lst);
-    }
-
-    const i = lst.indexOf(val);
-    if (i < 0) {
-        return exports.list_generator(lst);
-    }
-
-    const inners = [
-        exports.sub_list_generator(lst, i + 1, lst.length),
-        exports.sub_list_generator(lst, 0, i),
-    ];
-
-    return exports.chain(inners);
-};
-
-exports.reverse_wrap_exclude = function (lst, val) {
-    if (val === undefined) {
-        return exports.reverse_list_generator(lst);
-    }
-
-    const i = lst.indexOf(val);
-    if (i < 0) {
-        return exports.reverse_list_generator(lst);
+        }
     }
 
-    const inners = [
-        exports.reverse_sub_list_generator(lst, 0, i),
-        exports.reverse_sub_list_generator(lst, i + 1, lst.length),
-    ];
-
-    return exports.chain(inners);
-};
-
-exports.filter = function (gen, filter_func) {
-    return {
-        next() {
-            while (true) {
-                const val = gen.next();
-                if (val === undefined) {
-                    return undefined;
-                }
-                if (filter_func(val)) {
-                    return val;
-                }
-            }
-        },
-    };
-};
-
-exports.map = function (gen, map_func) {
-    return {
-        next() {
-            const val = gen.next();
-            if (val === undefined) {
-                return undefined;
+    for (let i = curr_stream_index + 1; i < streams.length; i += 1) {
+        const stream = streams[i];
+        for (const topic of get_topics(stream)) {
+            if (has_unread_messages(stream, topic)) {
+                return {stream, topic};
```

Comment:
@showell can you help review this commit?  And @andersk can you provide more detail in the commit message as to what's going on here?  There's a lot of code deleted and I'm not clear on whether this is just a reimplementation with the same effect or something else.

Reply:
(And also to what extent you believe the non-deleted tests validate the logic)

Reply:
The behavior should be the same; I certainly would have noted otherwise if not. I added “topic_generator previously included an entire lazy generator combinator library that was used four times. These straightforward equivalent loops might not be as fun but they are way simpler.”

The deleted tests were all for deleted code; the file retains 100% coverage.